### DISCUSSION TITLE
Why do we change rpcUrls structure in 0.4.x?

### DISCUSSION THREAD

**Discussion by npty** [wevm/wagmi#495] [Q&A]
I'm really like the concept of fallback mechanism. When some provider doesn't work, it switch to another provider, that's great. 

However, in version `0.4.x`, the `rpcUrls` structure is
```
rpcUrls: {
  alchemy?: string | undefined,
  infura?: string | undefined,
  default: string
}
```

The problem with this structure is: i cannot specify more than 1 provider if that chain hasn't supported by `alchemy` or `infura`. 

Why don't we just keep rpcUrls type to be `string[]`?

**Comment by tmm** [comment]
You can add other arbitrary key value pairs to the `rpcUrls` object. `rpcUrls` isn't related to the fallback mechanism.

**Reply by gordomium** [reply]
I have the same issue. Seems the interface is not allowed array here.

```ts
export declare type Chain = {
    // ...
    rpcUrls: {
        [key in RpcProviderName]?: string;
    } & {
        [key: string]: string;
        default: string;
    };
    // ...
};
```

Seems every wagmi wrapped connector only accept this kind of `Chain` type.

Did you mean working against the type? Or just my misunderstanding?

**Reply by tmm** [reply]
You can define any number of key value pairs for the `rpcUrls` object (in addition to the named `alchemy`,`infura`, and `default` keys):

```ts
const myChain: Chain = {
  id: 12345,
  name: 'My Chain',
  rpcUrls: {
    alchemy: 'https://alchemy.com/mychain',
    infura: 'https://infura.io/mychain,
    default: 'https://mychain.com',
    foo: 'https://foo.com/mychain',
    bar: 'https://bar.com/mychain',
  },
  ...
}
```

**Reply by gordomium** [reply]
Sorry, i still not get it.

As far as I know, there are two situations to do with `multi-rpcUrls`.


#### 1. Multi provider vendors

```ts
const myChain: Chain = {
  rpcUrls: {
    alchemy: 'https://alchemy.com/mychain',
    infura: 'https://infura.io/mychain,
    default: 'https://mychain.com',
    foo: 'https://foo.com/mychain',
    bar: 'https://bar.com/mychain',
  },
  ...
}
```

As your example, this can be used for customized providers.


#### 2. Fallback mechanism

```ts
const myChain: Chain = {
  rpcUrls: {
    default: [
       'https://node-1.mychain.com',
       'https://node-2.mychain.com',
       'wss://node-3.mychain.com',
    ],
  },
  ...
}
```

This is for preventing issues from bad network between clients end endpoints.

----

I just checkout the documents and code and did not find the `situation 2` solution.

**Reply by jxom** [reply]
Have you checked out https://wagmi.sh/docs/providers/configuring-chains? This covers both multiple providers & fallbacks.

**Reply by gordomium** [reply]
Yes. But that seems should use for named providers. Like switching between `alchemy`, `infura`, `public` ...

But it's not for some unnamed equivalence endpoints. I may have to do like this:

```ts
const chains: Omit<Chain>[] = {
  rpcUrls: {
    default: 'https://node0',
    customized_1:  'https://node1',
    customized_2:  'https://node2',
    // maybe ten more
  }
  // ...
}


const { chains, provider, webSocketProvider } = configureChains(chains, [
  publicProvider(), //  or jsonRpcProvider, They only consume `default`
  customizedConfigReadingProvider('customized_1'), 
  customizedConfigReadingProvider('customized_2'), 
  // ...
])
  
return createClient({
  connectors: [
    new Connector, // ...
  ],
  provider,
  webSocketProvider,
})

```

**Reply by jxom** [reply]
Yeah, you can use [multiple `jsonRpcProvider`'s](https://wagmi.sh/docs/providers/jsonRpc) for the custom RPC urls.

**Reply by gordomium** [reply]
Well... It looks like a workaround. I didn't aware that is by design.
Anyway.

Thank you for the help.
